X11: fix usage of required configuration values

AlwinEsch opened the pull request:
## Description
Before has Kodi checked only the **EGL_NATIVE_VISUAL_ID** and ignored
others. With this was e.g. on some addons the needed **EGL_DEPTH_SIZE**
not available.

This separates a part of `IsSuitableVisual(...)` into a separate function
(`SuitableCheck(...)`) which is checked by `GetEGLConfig(...)` and takes the
necessary configuration from the list.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
Sit for a while because the addons which use GL to switch to the 4.0 version so that they work again and was almost senseless.

Thought all the time to have done something wrong in the rework and now found that the error lies in Kodi.

In addition, I found in the end that it was with the override `KODI_GL_INTERFACE=GLX ./Kodi-x11` went and finally found that the error lies in Kodi itself.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

pkerling commented on xbmc/windowing/X11/GLContextEGL.cpp:
```diff
@@ -205,7 +205,7 @@ bool CGLContextEGL::CreatePB()
     EGL_GREEN_SIZE, 8,
     EGL_RED_SIZE, 8,
-    EGL_DEPTH_SIZE, 8,
+    EGL_DEPTH_SIZE, 24,
```
We request 16 bits depth buffer on most (all?) other platforms. Is that not enough? I'm not sure all (embedded) devices can support 24. @lrusak ?

AlwinEsch replied:
Not sure, seen only on all other places a used "24" (also a bit below in this file) and seen this "8" as c&p from line above.

AlwinEsch replied:
@FernetMenta better to change this and the "if (!eglGetConfigAttrib(eglDisplay, config, EGL_DEPTH_SIZE, &value) || value < 24)" to 16?

FernetMenta replied:
The pixel buffer, which is used for offscreen rendering, does not need this modification.

Regarding normal rendering, there is nothing in the Kodi application that requires a depth buffer. I won't waste resources just for the odd case of a fancy screen saver that does depth testing. Addons could request special features, like a depth buffer, on demand. That saves overhead not needed in general.

pkerling replied:
@FernetMenta In principle, you are of course right - Kodi does not need a depth buffer. But on-demand depth buffer requests will (at least I assume so) not go into v18 since the changes that would be required are too far-reaching. Right now, the behavior we have is a regression compared to v17. A number of visualizations and screensavers that used to work are now broken due to the depth buffer being absent.

As I don't think requesting the unused depth buffer will incur more overhead than the space it occupies in the GPU RAM and we were fine having it until now, I'd keep it for the time being. When we would make the depth buffer available on demand in the future, do you have any insight into how much the GLX/EGL surface recreation would const, i.e. would it possibly stutter for example?

As a side-note, `GLContextGLX` also requests 24 bits of depth buffer currently.

pkerling commented:
@AlwinEsch Can you make the changes? IMO this should go into v18 (@MartijnKaijser) so we have working viz and screen savers on X11.

AlwinEsch commented:
Is updated. 

Sit still to change the addons, unbelievably complex to convert these to the shader type :-| .

pkerling commented:
@FernetMenta Can you comment on why there is this custom visual selection code in `GetEGLConfig` (that - strictly speaking - misses some checks such as `EGL_RENDERABLE_TYPE` and `EGL_SURFACE_TYPE`) as opposed to using `eglChooseConfig` like in the PB code?

FernetMenta commented:
@pkerling when selecting a config for rendering to the main window GetEGLConfig returns a config that matches the native visual. Such a config has all the attributes required by Kodi core.

pkerling commented:
> when selecting a config for rendering to the main window GetEGLConfig returns a config that matches the native visual. Such a config has all the attributes required by Kodi core.

I believe you that it does (after all it is working fine so far), but that does not really answer my question :-) Which is why `eglChooseConfig` is basically reimplemented here but without the priority logic and ignoring important bits such as `EGL_RENDERABLE_TYPE` and `EGL_SURFACE_TYPE`. Theoretically, you might get a config that is only suitable for GLES. And if the first config in the list that has the right visual does not fit the other criteria, we currently just bail out instead of looking for a fitting one. Fortunately, I think you will usually get a window-surface-/GL-capable RGBA8888 config first so it hasn't been a problem.

pkerling commented on xbmc/windowing/X11/GLContextEGL.cpp:
```diff
@@ -334,18 +334,25 @@ void CGLContextEGL::Detach()
 bool CGLContextEGL::IsSuitableVisual(XVisualInfo *vInfo)
 {
   EGLConfig config = GetEGLConfig(m_eglDisplay, vInfo);
+  return SuitableCheck(m_eglDisplay, config);
```
`GetEGLConfig` now calls `SuitableCheck`, so this call is superfluous, making the whole `IsSuitableVisual` function superfluous since it then does nothing but call `GetEGLConfig`. I think you can remove it.

AlwinEsch replied:
Is removed and changed to "return config != EGL_NO_CONFIG;".

AlwinEsch commented:
I had to add a second commit. When switching I noticed another error.

Some addons need the stuff described above, but they have always been removed by Kodi with glClear.

AlwinEsch commented:
Here the difference between of second commit:

Before:
![bildschirmfoto von 2019-01-10 18-40-52](https://user-images.githubusercontent.com/6879739/50986694-c5437e80-1507-11e9-8231-f93eda4fe84d.png)

After:
![bildschirmfoto von 2019-01-10 18-42-02](https://user-images.githubusercontent.com/6879739/50986740-eb691e80-1507-11e9-87df-5cc07fe5b874.png)

pkerling commented on xbmc/windowing/X11/GLContextEGL.h:
```diff
@@ -34,6 +34,7 @@ class CGLContextEGL : public CGLContext
   EGLConfig m_eglConfig;
 protected:
   bool IsSuitableVisual(XVisualInfo *vInfo);
+  bool SuitableCheck(EGLDisplay& eglDisplay, EGLConfig& config);
```
```suggestion
  bool SuitableCheck(EGLDisplay eglDisplay, EGLConfig config);
```

Both are pointers, no need for a reference

pkerling commented on xbmc/windowing/X11/GLContextEGL.cpp:
```diff
@@ -334,18 +334,25 @@ void CGLContextEGL::Detach()
+  return config != EGL_NO_CONFIG;
+}
+
+bool CGLContextEGL::SuitableCheck(EGLDisplay& eglDisplay, EGLConfig& config)
```
```suggestion
bool CGLContextEGL::SuitableCheck(EGLDisplay eglDisplay, EGLConfig config)
```

pkerling commented on xbmc/windowing/X11/GLContextEGL.cpp:
```diff
@@ -334,18 +334,25 @@ void CGLContextEGL::Detach()
+
+bool CGLContextEGL::SuitableCheck(EGLDisplay& eglDisplay, EGLConfig& config)
+{
   if (config == EGL_NO_CONFIG)
```
You can move this block to `IsSuitableVisual`

AlwinEsch replied:
Your mean to let the
```cpp
  if (config == EGL_NO_CONFIG)
  {
    CLog::Log(LOGERROR, "Failed to determine egl config for visual info");
    return false;
  }
```
away?

AlwinEsch replied:
And:
```cpp
bool CGLContextEGL::IsSuitableVisual(XVisualInfo *vInfo)
{
  EGLConfig config = GetEGLConfig(m_eglDisplay, vInfo);
  if (config == EGL_NO_CONFIG)
  {
    CLog::Log(LOGERROR, "Failed to determine egl config for visual info");
    return false;
  }
  return true;
}
```

AlwinEsch replied:
If yes makes not direct sense, before was it only to report a failed GetEGLConfig and to prevent the further check. But now becomes everything done inside the GetEGLConfig and can report EGL_NO_CONFIG frequently.

pkerling replied:
Why should `GetEGLConfig` frequently report `EGL_NO_CONFIG`? In that case we will not get any display

AlwinEsch replied:
I have removed the IsSuitableVisual call complete, where this becomes called is the GetEGLConfig called directly after. 

Now think that it is a remnant of the conversion of the old CGLContextGLX where this call was necessary.

The ```CLog::Log(LOGERROR, "Failed to determine egl config for visual info");``` should never come, only for the worst case where a "eglGetConfigs" call returns true with a empty entry in list.

yol reviewed (commented):
Few minors then I think we are good to go

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
How is the state to bring in? The most OpenGL related addons need this.

pkerling commented on xbmc/windows/GUIWindowScreensaver.cpp:
```diff
@@ -42,6 +42,21 @@ void CGUIWindowScreensaver::Render()
+    if (m_renderClearCnt == 0)
+      return;
+
+    /* make five render calls with clear to prevent mix with cached display data */
```
???

AlwinEsch replied:
Yeah that's really crazy, but without that, when you start some visualizations (which used previous screen data and no "glClear") it flickers to the wild (for example, some modes in Screensaver Solar Winds)

pkerling replied:
This hack should under no circumstances go in.

I just noticed another thing: The screensavers render directly to display, right? If so, not calling glClear will not solve the problem for example on Wayland. We call `eglSwapBuffers` to display the content, but this is not a simple "front-back-buffer-switch" which it might have been in former times. Modern graphics drivers have multiple buffers ready and cycle through them as they see fit. Spec says: "The contents of ancillary buffers are always undefined after calling eglSwapBuffers. The contents of the color buffer are undefined if the value of the EGL_SWAP_BEHAVIOR attribute of surface is not EGL_BUFFER_PRESERVED." So the color buffer content is undefined when not calling `glClear` and the screensaver cannot rely on any specific content still being there.

AlwinEsch committed:
X11: remove not needed CGLContextEGL::IsSuitableVisual

This calls GetEGLConfig which becomes also called direct after
the place where IsSuitableVisual is called.

AlwinEsch commented on xbmc/windows/GUIWindowScreensaver.cpp:
```diff
@@ -68,6 +68,9 @@ bool CGUIWindowScreensaver::OnMessage(CGUIMessage& message)
   case GUI_MSG_WINDOW_INIT:
     {
+      // override the clear colour - we must never clear fullscreen
+      m_clearBackground = 0;
```
Found another way, the way before with count was really not nice.

There are also a few other places in Kodi that go that way. Also, 3D video rendering must avoid a glClear (again other way) is called.

pkerling replied:
See https://github.com/xbmc/xbmc/pull/15148#discussion_r252569744 - not calling `glClear` does not mean that the buffer contents are preserved.

pkerling replied:
> Why is it then also needed on other fullscreen places in Kodi?

How do you know it is needed? It might just be a performance optimization (no use clearing the buffer if we draw over everything anyway) - do you have a specific place where you know that it is really necessary?

> Does the complete screen content have to be copied with every render call to have it back at the next frame?

Yes and no. There are extensions that can be used instead of the often unsupported `EGL_BUFFER_PRESERVED`. But they will not give you the last frame, but instead an arbitrary previous frame and tell you how old it is (`EGL_EXT_buffer_age`) - this also works with Wayland. But it will not help with the solarwinds screensaver because I guess that it is implemented in a way that it needs the exact previous frame. Then yes, you need to copy when swapping buffers. This is also what the driver would usually do for you anyway when setting `EGL_BUFFER_PRESERVED`. See also e.g. https://community.arm.com/graphics/b/blog/posts/mali-performance-3-is-egl_5f00_buffer_5f00_preserved-a-good-thing

Other option is to modify solarwinds screensaver to always draw everything.

AlwinEsch commented:
Have now also tested with wayland over https://github.com/notspiff/screensavers.rsxs/pull/20 and runs without problems

AlwinEsch committed:
X11: fix usage of required configuration values

Before has Kodi checked only the EGL_NATIVE_VISUAL_ID and ignored
others. With this was e.g. on a some addons the needed EGL_DEPTH_SIZE
not available.

This separates a part of IsSuitableVisual (...) into a separate function
(SuitableCheck (...)) which is checked by GetEGLConfig and takes the
necessary configuration from the list.

AlwinEsch commented:
Have removed the last commit. Must be for Leia now a separate request created?

yol reviewed (approved):
Should be OK now

pkerling commented:
Afaik we have not branched yet, so no backport needed

pkerling commented:
@AlwinEsch thanks for your persistence! happy to see the screensavers and visualizations get some love :)
